docker/ubuntu/Dockerfile

```diff
@@ -1,138 +1,166 @@
-    cd .. && \
-    rm -rf vdatum
-
-## compile and install PDAL (not available on Debian/Ubuntu) with laz-perf enabled
-ENV NUMTHREADS=4
-WORKDIR /src
-RUN wget -q \
- https://github.com/PDAL/PDAL/releases/download/${PDAL_VERSION}/PDAL-${PDAL_VERSION}-src.tar.gz && \
-    tar xfz PDAL-${PDAL_VERSION}-src.tar.gz && \
-    cd /src/PDAL-${PDAL_VERSION}-src && \
-    mkdir build && \
-    cd build && \
-    cmake .. \
-      -G "Unix Makefiles" \
-      -DCMAKE_BUILD_TYPE=Release \
-      -DCMAKE_INSTALL_PREFIX=/usr \
-      -DCMAKE_C_COMPILER=gcc \
-      -DCMAKE_CXX_COMPILER=g++ \
-      -DCMAKE_MAKE_PROGRAM=make \
-      -DBUILD_PLUGIN_PYTHON=ON \
-      -DBUILD_PLUGIN_CPD=OFF \
-      -DBUILD_PLUGIN_GREYHOUND=ON \
-      -DBUILD_PLUGIN_HEXBIN=ON \
-      -DHEXER_INCLUDE_DIR=/usr/include/ \
-      -DBUILD_PLUGIN_NITF=OFF \
-      -DBUILD_PLUGIN_ICEBRIDGE=ON \
-      -DBUILD_PLUGIN_PGPOINTCLOUD=OFF \
-      -DBUILD_PGPOINTCLOUD_TESTS=OFF \
-      -DBUILD_PLUGIN_SQLITE=ON \
-      -DWITH_LASZIP=OFF \
-      -DWITH_LAZPERF=ON \
-      -DWITH_TESTS=ON && \
-    make -j $NUMTHREADS && \
-    make install
+# Start build stage
+FROM common as build
+
+# Define build packages
+ENV GRASS_BUILD_PACKAGES="\
+    cmake \
+    libbz2-dev \
+    libcairo2-dev \
+    libopenblas-dev \
+    libpdal-dev \
+    libproj-dev \
+    libpq-dev \
+    libfftw3-dev \
+    libfreetype6-dev \
+    libomp-dev \
+    zlib1g-dev \
+    libnetcdf-dev \
+    libopenjp2-7-dev \
+    libreadline-dev \
+    libjpeg-dev \
+    libpnglite-dev \
+    libsqlite3-dev \
+    libtiff-dev \
+    libzstd-dev \
+    libncurses5-dev \
+    mesa-common-dev \
+    zlib1g-dev \
+    "
+
+# Add build packages
+RUN apt-get update \
+    && apt-get install  -y --no-install-recommends --no-install-suggests \
+    $GRASS_BUILD_PACKAGES \
+    && apt-get clean all \
+    && rm -rf /var/lib/apt/lists/*
+
+# Add pip and Python packages
+RUN (echo "Install Python" \
+    && wget https://bootstrap.pypa.io/pip/get-pip.py \
+    # && apt-get install -y python3-ensurepip \
+    && python3 get-pip.py \
+    # && python3 -m ensurepip --upgrade \
+    && rm -r get-pip.py \
+    && mkdir -p /src/site-packages \
+    && cd /src \
+    && python3 -m pip install --no-cache-dir -t /src/site-packages --upgrade \
+    pip \
+    setuptools \
+    grass-session \
+    python-dateutil \
+    python-magic \
+    numpy \
+    Pillow \
+    ply \
+    setuptools \
+    matplotlib \
+    psycopg2 \
+    && rm -r /root/.cache \
+    )
+
+# ers are encouraged to fetch vertical datums and store into PROJ dir, e.g.:
+# WORKDIR /src
+# RUN mkdir vdatum && \
+#     cd vdatum && \
+#     wget -q http://download.osgeo.org/proj/vdatum/usa_geoid2012.zip && unzip -j -u usa_geoid2012.zip -d /usr/share/proj; \
+#     cd .. && \
+#     rm -rf vdatum
 
 # copy grass gis source
-WORKDIR /src
 COPY . /src/grass_build/
+
 WORKDIR /src/grass_build
 
 # Cleanup potentially leftover GISRC file with wrong path to "demolocation"
```

Comment:
On the Line after (168, I can't comment directly on it), there is a `RUN rm` line. This will create a new layer, to remove files in the build stage image, but the files should be cleaned in the same step as they are created to not waste space. This is correct if the files to remove were added from the COPY step. In that case, maybe consider using a .dockerignore file (https://docs.docker.com/engine/reference/builder/#dockerignore-file) to not have them transferred in the first place.

Since this rm is not in the final stage, it would not impact the image produced at the end.